crates/accelerate/src/synthesis/evolution/mod.rs

```diff
@@ -0,0 +1,336 @@
+    for (i, q) in qubits.iter().enumerate() {
+        v[*q as usize] = sparse_pauli.chars().nth(i).unwrap();
+    }
+    v.into_iter().collect()
+}
+
+/// Return the Qiskit's gate corresponding to the given Rustiq's Clifford gate.
+fn qiskit_clifford_gate(rustiq_gate: &CliffordGate) -> QiskitGate {
+    match rustiq_gate {
+        CliffordGate::CNOT(i, j) => (
+            StandardGate::CXGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32), Qubit(*j as u32)],
+        ),
+        CliffordGate::CZ(i, j) => (
+            StandardGate::CZGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32), Qubit(*j as u32)],
+        ),
+        CliffordGate::H(i) => (
+            StandardGate::HGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32)],
+        ),
+        CliffordGate::S(i) => (
+            StandardGate::SGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32)],
+        ),
+        CliffordGate::Sd(i) => (
+            StandardGate::SdgGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32)],
+        ),
+        CliffordGate::SqrtX(i) => (
+            StandardGate::SXGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32)],
+        ),
+        CliffordGate::SqrtXd(i) => (
+            StandardGate::SXdgGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32)],
+        ),
+    }
+}
+
+/// Return the Qiskit rotation gate corresponding to the single-qubit Pauli rotation.
+/// # Arguments
+///
+/// * py: a GIL handle, needed to negate rotation parameters in Python space.
+/// * paulis: Rustiq's data structure storing pauli rotations.
+/// * i: index of the single-qubit Pauli rotation.
+/// * angle: - Qiskit's rotation angle.
+fn qiskit_rotation_gate(py: Python, paulis: &PauliSet, i: usize, angle: &Param) -> QiskitGate {
+    let (phase, pauli_str) = paulis.get(i);
+    for (q, c) in pauli_str.chars().enumerate() {
+        if c != 'I' {
+            let standard_gate = match c {
+                'X' => StandardGate::RXGate,
+                'Y' => StandardGate::RYGate,
+                'Z' => StandardGate::RZGate,
+                _ => panic!(),
+            };
+            // we need to multiply the angle by 2
+            // we also need to negate it when there is a phase
+            let param = match phase {
+                false => multiply_param(angle, 2.0, py),
+                true => multiply_param(angle, -2.0, py),
+            };
+            return (standard_gate, smallvec![param], smallvec![Qubit(q as u32)]);
+        }
+    }
+    unreachable!()
+}
+
+/// A DAG that stores ordered Paulis, up to commutativity.
+struct CommutativityDag {
+    /// Rustworkx's DAG
+    dag: StableDiGraph<usize, ()>,
+}
+
+impl CommutativityDag {
+    /// Construct a DAG based on the commutativity relations between paulis.
+    pub fn from_paulis(paulis: &PauliSet) -> Self {
```

Comment:
Same question here, do these methods need to be public? Since the struct itself is not I would assume that we can just have these private? 🤔

Reply:
yes, fixed